pkg/guacanalytics/toposort.go

```diff
@@ -22,58 +22,65 @@ import (
 	"github.com/Khan/genqlient/graphql"
 )
 
-// TODO: add tests
-func ToposortFromBfsNodeMap(ctx context.Context, gqlClient graphql.Client, nodeMap map[string]BfsNode) (map[int][]string, []string, error) {
-	frontiers := make(map[int][]string)
-	parentsMap, infoNodes := copyParents(nodeMap)
-	frontierLevel := 0
+// TopoSortFromBfsNodeMap sorts the nodes such that it returns a map of level -> list of nodeIDs at that level
+func TopoSortFromBfsNodeMap(ctx context.Context, gqlClient graphql.Client, nodeMap map[string]BfsNode) (map[int][]string, []string, error) {
+	sortedNodes := make(map[int][]string) // map of level -> list of nodeIDs at that level
+	parentsMap, childrensMap, infoNodes := copyParents(nodeMap)
+	// parentsMap: map of nodeID (child) -> list of parents in the form of a map
+	// childrensMap: map of nodeID (parent) -> list of children in the form of a list
+	bfsLevel := 0
 	numNodes := 0
 	totalNodes := len(parentsMap)
 
 	for numNodes < totalNodes {
 		foundIDs := make(map[string]bool)
-		for id, parentsList := range parentsMap {
-			if len(parentsList) == 0 || (parentsList[0] == "" && len(parentsList) == 1) {
-				frontiers[frontierLevel] = append(frontiers[frontierLevel], id)
-				foundIDs[id] = true
+		for id, pMap := range parentsMap {
+			if pMap.parents != nil && len(pMap.parents) == 0 { // if this node has no parents, it is a root node
+				sortedNodes[bfsLevel] = append(sortedNodes[bfsLevel], id)
 				numNodes++
+				foundIDs[id] = true
+				delete(parentsMap, id)
 			}
 		}
 
-		if len(foundIDs) == 0 {
-			// TODO: print out offending cycle
-			return frontiers, infoNodes, fmt.Errorf("error: cycle detected")
-		}
-
 		for id := range foundIDs {
-			delete(parentsMap, id)
-		}
-
-		for id, parentsList := range parentsMap {
-			newParentsList := []string{}
-			for _, parentID := range parentsList {
-				if !foundIDs[parentID] {
-					newParentsList = append(newParentsList, parentID)
-				}
+			for _, childID := range childrensMap[id] { // loop through all the children of this node
+				delete(parentsMap[childID].parents, id) // remove this node from the map of parents of the child
 			}
+		}
 
-			parentsMap[id] = newParentsList
+		if len(foundIDs) == 0 {
```

Comment:
Can you add back in the TODO comment please: "// TODO: print out offending cycle"

Reply:
Added